### ISSUE TITLE
[GRPO] How can I train a PeftModel using grpo?

### ISSUE THREAD

**Issue by JinyuanSun** [huggingface/trl#3031]
Hi, I used lora for fine-tuning the pre-trained model. I want to apply grpo to the peftmodel with adapter, how can I do it?
This is my current code:
```python
dataset = load_dataset("trl-lib/tldr", split="train")

# Define the reward function, which rewards completions that are close to 20 characters
def reward_len(completions, **kwargs):
    return [-abs(20 - len(completion)) for completion in completions]

pre_trained_model = AutoModelForCausalLM.from_pretrained("Qwen/Qwen2-0.5B-Instruct")
adapter_path = '/path/to/saved_adapter'
model = PeftModel.from_pretrained(pre_trained_model, adapter_path)


training_args = GRPOConfig(output_dir="Qwen2-0.5B-GRPO", logging_steps=10)
trainer = GRPOTrainer(
    model=model,
    reward_funcs=reward_len,
    args=training_args,
    train_dataset=dataset,
)
trainer.train()
```
The reward is not increasing with the adatper loaded, how can I solve this?

**Comment by AndreiCComan** [comment]
@JinyuanSun I had a similar issue in #2856 which has been fixed. Could you try to run the same MRE and the latest changes (i.e., learning rate etc.) I posted there?

**Comment by zhangwengyu999** [comment]
Same issue.

#2856 is not the same issue, since we want to perform GRPO on a fine-tuned PeftModel, but not perform GRPO together with PEFT.

**Comment by DingZhenChen-code** [comment]
Same issue. How to continue training on a fine-tuned PeftModel?

**Comment by cliang-huanglab** [comment]
Same issue. Have you found a solution?

**Comment by EttoreCaputo** [comment]
@JinyuanSun After many attempts, this worked for me:
```python
# LOAD THE BASE MODEL AND APPLY THE PREVIOUS FINE-TUNED (SFT in my case) PEFT MODEL
model, tokenizer = FastLanguageModel.from_pretrained(
    model_name = "BASE MODEL HERE (unsloth in my case)",
    max_seq_length = max_seq_length,
    load_in_4bit = True,
    fast_inference = True,
    max_lora_rank = lora_rank,
    gpu_memory_utilization = 0.5,
)

model = PeftModel.from_pretrained(
    model, 
    "ADAPTER PATH HERE",
    is_trainable = True,
)

FastLanguageModel.patch_peft_model(model)
```


```python
#GRPO TRAIN
training_args = GRPOConfig(
    use_vllm = True,
    ... #other params
)

trainer = GRPOTrainer(
    model = model,
    processing_class = tokenizer,
    reward_funcs = [
        ...#reward functions
    ],
    args = training_args,
    train_dataset = dataset,
)
trainer.train()
```

**Comment by albertvillanova** [comment]
Hi @JinyuanSun, sorry for the very late reply — we're catching up on the backlog. 

This isn't a TRL bug: `PeftModel.from_pretrained(base, adapter_path)` loads the adapter in inference mode (`is_trainable=False`) by default, so every LoRA parameter has `requires_grad=False`. GRPO then has *zero* trainable parameters and the optimizer changes nothing, which is exactly why your reward stays flat with the adapter loaded but rises when you train a fresh one. 

The fix is to load it as trainable:
```python
model = PeftModel.from_pretrained(pre_trained_model, adapter_path, is_trainable=True)
trainer = GRPOTrainer(model=model, reward_funcs=reward_len, args=training_args, train_dataset=dataset)
```

Current TRL also auto-handles the reference model for a pre-loaded PeftModel, so you don't need to pass `peft_config` again (see https://github.com/huggingface/trl/blob/0ed4ac7fa03bf388a94566f1b1f71e71445ca6e7/trl/trainer/grpo_trainer.py#L368-L376).

Note this is distinct from #2856, which is about PEFT-from-scratch + vLLM. 

Closing as not-a-bug, but happy to re-open if the reward still doesn't move after setting `is_trainable=True`.
